src/lib.rs

```diff
@@ -130,23 +199,44 @@ where
 {
     fn on_new_span(&self, attr: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
         let span = ctx.span(id).expect("Span not found, this is a bug");
-        let parent = span.parent();
-        let mut extensions = span.extensions_mut();
+        // try to identify a `trace_id` for this span
+        let trace_id = if let Some(trace_id) = trace_id::from_span(&span, attr) {
+            trace_id
+        } else {
+            // if none can be found, don't emit a Segment
+            return;
+        };
 
-        let segment = model::Segment {
-            name: {
-                // What the docs say:
-                // The logical name of the service that handled the request, up
-                // to 200 characters. Names can contain Unicode letters,
-                // numbers, and whitespace, and the following symbols: _, ., :,
-                // /, %, &, #, =, +, \, -, @.
+        let record = Record::new(attr.values());
+        let (mut metadata, mut annotations) = crate::model::metadata_and_annotations_from(&record);
 
-                // What we do:
-                // Use the static name ascribed by the user to the span.
+        let kind = match attr.fields().field(TRACE_ID_FIELD).is_some() {
+            true => model::Kind::Segment,
+            false => model::Kind::Subsegment,
+        };
+
+        annotations
+            .fields
+            .add("tracing.target", attr.metadata().target());
+        metadata.fields.add("tracing.file", attr.metadata().file());
+        metadata.fields.add("tracing.line", attr.metadata().line());
+
+        // if we're creating a segment, retain the tracing span name as an annotation
+        // otherwise, it'll be available as the name of the subsegment.
+        if let model::Kind::Segment = kind {
+            annotations
+                .fields
+                .add("tracing.name", attr.metadata().name());
```

Comment:
can we make this a constant?

Reply:
I'm a little hesitant to make a change that increases indirection without decreasing repetition (since these are the only places these string literals occur).

Reply:
ok